### DISCUSSION TITLE
How to backup all data on netbox?

### DISCUSSION THREAD

**Discussion by ssangho8** [netbox-community/netbox#9681] [Q&A]
Hi, 

I'm operating netbox through running docker-compose.

I'm wondering if there's a way to backup all data (e.g. IP addresses, Devices) on netbox.

 I want to backup data every day or every week.

Here's the information when I type "docker ps -a" command on my netbox server.

# docker ps -a
CONTAINER ID   IMAGE                               COMMAND                  CREATED        STATUS        PORTS                                       NAMES
b3379a749353   netboxcommunity/netbox:v3.2-1.6.1   "/sbin/tini -- /opt/…"   2 months ago   Up 2 months   0.0.0.0:8000->8080/tcp, :::8000->8080/tcp   netboxdocker_netbox_1
bcd810b00e02   netboxcommunity/netbox:v3.2-1.6.1   "/sbin/tini -- /opt/…"   2 months ago   Up 2 months                                               netboxdocker_netbox-housekeeping_1
fc17e44536a7   netboxcommunity/netbox:v3.2-1.6.1   "/sbin/tini -- /opt/…"   2 months ago   Up 2 months                                               netboxdocker_netbox-worker_1
c3eb6d82b84e   postgres:14-alpine                  "docker-entrypoint.s…"   2 months ago   Up 2 months   5432/tcp                                    netboxdocker_postgres_1
b16e3fb413e5   redis:6-alpine                      "docker-entrypoint.s…"   2 months ago   Up 2 months   6379/tcp                                    netboxdocker_redis-cache_1
feb69922effa   redis:6-alpine                      "docker-entrypoint.s…"   2 months ago   Up 2 months   6379/tcp                                    netboxdocker_redis_1

Please help.

**Comment by PieterL75** [comment]
Use the postgres pg_dump to backup the database :
https://docs.netbox.dev/en/stable/administration/replicating-netbox/

**Reply by Nino-Rey** [reply]
I'm also thinking the same way Netbox Devices has Export function which return a CSV file and I'd like to run it weekly to backup all our Devices and put it on S3 bucket. Is there any reference that I could look at in this repo?

**Reply by candlerb** [reply]
That won't work, because you cannot restore Netbox from CSV files.  Firstly, in general CSV files don't include unique ID references for one object to refer to another.  And secondly, there are circular references which have to be restored simultaneously (in particular: a device's primary_ip points to an IP address, but the IP address points to an interface which points to the device).

Therefore, you'll will have to use pg_dump to make a usable Netbox backup.  Copying it to an S3 bucket can be done in a script using awscli or s3cmd.

**Comment by ssangho8** [comment]
I installed psql v10 and I get error message below when I type "psql" on CLI.

psql: could not connect to server: No such file or directory
        Is the server running locally and accepting
        connections on Unix domain socket "/var/run/postgresql/.s.PGSQL.5432"?

Should I manage psql within the docker container? or on local server?

**Reply by candlerb** [reply]
Easiest is to enter the postgres container and run `psql` inside the container (i.e. use `docker exec ...`)

In principle you can install the postgres client on your server or a remote workstation, and connect to the postgres database over the network. However this may require some setting up, for example adding rules to `hba.conf` to allow connections from your workstation's IP address.

**Reply by ssangho8** [reply]
I currently can't find where hba.conf file located at.

I just got in to the 'netbox' database on postgresql container.

**Reply by candlerb** [reply]
You'll need to switch your userid to the user that postgres runs under, which if the container is based on Ubuntu will probably be user `psql`:

```
sudo -u postgres psql netbox
```

For things like hba.conf, please refer to postgres documentation or a postgres user group.

**Reply by ssangho8** [reply]
In postgres.env file located at /env, 
POSTGRES_DB=netbox
POSTGRES_PASSWORD=~~
POSTGRES_USER=netbox

I got into the postgres container after running below.
#docker exec -it --user root netboxdocker_postgres_1 /bin/bash

Also I could get into the netbox database after running below.
bash-5.1# psql netbox -U netbox

Do you mean I have to use pg_dump so that I could export whole data on netbox database?

**Reply by candlerb** [reply]
Yes.  You'll be able to combine it into a single command, probably some variation of the following (untested as I don't run netbox in docker):

```
docker exec --user root netboxdocker_postgres_1 /usr/local/bin/pg_dump netbox -U netbox
```

**Comment by mingueladavid** [comment]
This is my first time using netbox and postgres. After a lot of test try to find the correct data base and restore procedure I created the following script to generate the data base backup. It only work due to permission problem in /var/lib/postgresql/ and  the restore should be do using postgres user. I used the as part of the backup the option --clean, because if you restore over the actual database it give you errors "tables already exist" sorry for my English because my firts languaje is spanish.  The owner of this script should be postgres. 

#!/bin/sh

# **********************************************************************************************************
# 05-jul-2023 DRM
# Script para correr backup de la base de datos de Netbox. Este script debe estar localizado
# y debe correrse en el directorio /var/lib/postgresql. El script debe tener como  ownership a postgres
# -rwxr--r-- 1 postgres postgres  db_backup.sh
# La restauración de la base de datos debe hacerlo bajo el usuario de postgres
# us - postgres
# postgres@netboxU20:~$ psql -e -d netbox < netbox-mysql-FECHA_DEL_BACKUP.sql
# El backup de la base de datos se guardara los ultimos 5 dias.
# **********************************************************************************************************

N="netbox-mysql"
D=`date +"%Y%m%d"`
Z=`date +"%Y%m%d" -d "-5 days"`

BKP="/home/oberon"
ARC="/var/backup"

/usr/bin/su -c  'pg_dump -U postgres -F p --clean netbox > /var/lib/postgresql/db_clean.sql' postgres
/usr/bin/mv /var/lib/postgresql/db_clean.sql /var/lib/postgresql/$N-$D.sql
/usr/bin/rm /var/lib/postgresql/$N-$Z.sql

**Reply by candlerb** [reply]
> N="netbox-mysql"

Your colleagues may find that confusing, since this is a dump from postgres not mysql :-)

This is my rough-and-ready backup script:

```
#!/bin/sh -e
mkdir -p /var/backups/netbox
chmod 750 /var/backups/netbox
su -c 'pg_dump --exclude-table-data extras_cachedvalue --exclude-table-data extras_objectchange netbox' postgres |
  gzip -9 >/var/backups/netbox/netbox.$(date +%Y%m%d%H%M).gz
```

called from a cronjob:

```
#!/bin/sh -eu
/usr/local/bin/netbox-backup
find /var/backups/netbox -mtime +90 -type f -delete
```

**Reply by mingueladavid** [reply]
Thanks, yes you right N="netbox-mysql" but is just a variable......

**Comment by PieterL75** [comment]
I have this : 

```
#!/bin/bash

cd /opt/backup/netbox

DATETIME=$(date +%y%m%d_%H%M)
MAXDAYS=45

#database
pg_dump -Fc --username netbox --host localhost netbox > $(hostname)_$DATETIME.sql
find . -name "*.sql" -maxdepth 1 -mtime +$MAXDAYS -delete

#files
tar -czf $(hostname)_$DATETIME.files.tgz --exclude=".git" --exclude="__pycache__" /opt/netbox/netbox/local /opt/netbox/netbox/media /opt/netbox/netbox/reports /opt/netbox/netbox/scripts
find . -name "*.files.tgz" -maxdepth 1 -mtime +$MAXDAYS -delete

#configuration
crontab -l >/opt/netbox/netbox/netbox/netbox.crontab
tar -czf $(hostname)_$DATETIME.configs.tgz --exclude=".git" --exclude="__pycache__" /opt/netbox/local_requirements.txt /opt/netbox/netbox/netbox/configuration.py /etc/nginx/sites-available/netbox /opt/netbox/netbox/netbox/netbox.crontab
find . -name "*.configs.tgz" -maxdepth 1 -mtime +$MAXDAYS -delete

sftp mybackupserver:/opt/backup/netbox <<< $"put $(hostname)_$DATETIME.sql"
sftp mybackupserver:/opt/backup/netbox <<< $"put $(hostname)_$DATETIME.files.tgz"
sftp mybackupserver:/opt/backup/netbox <<< $"put $(hostname)_$DATETIME.configs.tgz"

/usr/bin/find /opt/backup/netbox -maxdepth 1 -mtime +5 -name "netbox*" -print -delete
```

and in the /home/netbox folder I have
/home/netbox/.pgpass
with
```
localhost:*:netbox:netbox:mysecretpassword
```

all ran from the 'netbox' crontab (3 times a day)
```
10 02 * * * /home/netbox/nbbackup.sh
10 12 * * * /home/netbox/nbbackup.sh
10 18 * * * /home/netbox/nbbackup.sh
```

**Reply by Lord-ZuzurNC** [reply]
Hello,

I'm a gravedigger but I'm wondering why you delete 45+ days old sql files in line 10 then all 5+ days old file starting by netbox (so sql as well if hostname start with that word, which may happens for a lot of us) at line 25 (last).

(less important but why using "find" then "/usr/bin/find" at the end?)

**Reply by Lord-ZuzurNC** [reply]
I rewrote the script to suit me (using @PieterL75 and @candlerb comments). If someone does have similar needs, here you go bellow.
I added a "BAKDIR" variable, the backup of the configuration file "gunicorn.py".
I deleted the sftp commands (as I will manage using another method) and the last deleting 5+ days old files.
```
#!/bin/bash

BAKDIR=/opt/backup
DATETIME=$(date +%y%m%d_%H%M)
MAXDAYS=31

cd $BAKDIR

# PostgreSQL Database Backup
pg_dump -Fc --exclude-table-data extras_cachedvalue --exclude-table-data extras_objectchange --username netbox --host localhost netbox > ./$(hostname)_$DATETIME.sql
find $BAKDIR -name "*.sql" -maxdepth 1 -mtime +$MAXDAYS -delete

# Data Backup
tar -czf $(hostname)_$DATETIME.files.tgz --exclude=".git" --exclude="__pycache__" /opt/netbox/netbox/media /opt/netbox/netbox/reports /opt/netbox/netbox/scripts
find $BAKDIR -name "*.files.tgz" -maxdepth 1 -mtime +$MAXDAYS -delete

# Configuration Backup
crontab -l > /opt/netbox/netbox/netbox/netbox.crontab
tar -czf $(hostname)_$DATETIME.configs.tgz --exclude=".git" --exclude="__pycache__" /opt/netbox/local_requirements.txt /opt/netbox/netbox/netbox/configuration.py /opt/netbox/gunicorn.py /etc/nginx/sites-available/netbox /opt/netbox/netbox/netbox/netbox.crontab
find $BAKDIR -name "*.configs.tgz" -maxdepth 1 -mtime +$MAXDAYS -delete
```

I also add, if anyone come here and to the same, that you need to change the permission to 0600 for the .pgpass
`chmod 0600 ~/.pgpass`

**Comment by FlorianHeigl** [comment]
my take, please feel free to merge into yours to get something out of it.
[backup-netbox-docker-db.sh](https://gist.github.com/FlorianHeigl/da9a4e9090f31402523cde3cb81312ac#file-backup-netbox-docker-db-sh)

**Comment by diasdmhub** [comment]
Isn't there a way to export all the data in a YAML or JSON format and easily import it back?

**Reply by Lord-ZuzurNC** [reply]
I don't think you can have a single big document so you'll need to do one per category. Like one for devices, one for vlans, one for IP subnets etc...

I think exportation order doesn't matter but importation order will be important as you'll not be able to import, for example, vlan with associated IP subnet if you didn't import the IP subnet beforehand.

You will either need a script that calls the API or a tool like Bruno and a collection to export and another one to import that you can play manually.

**Comment by diasdmhub** [comment]
The thing is that, when I export a category, the exported default CSV columns do not match the imported fields.

For example, this is the result of a sample _Tenents_ export.

```csv
Name,Group,Description,ID,Slug,Comments,Contacts,Tags,Created,Last updated
Apt,,,1,apt,,,cas,2025-05-03T00:50:24.143437+00:00,2025-05-03T01:12:28.076655+00:00
Cond,,,2,cond,,,ie,2025-05-03T00:50:40.435945+00:00,2025-05-03T01:12:34.678913+00:00
```

Netbox shows the _Tenent_ import objects as:

![tenent_fields](https://github.com/user-attachments/assets/c0e09c4f-dee1-4421-a952-0795ad71e96b)

When I try to import the default CSV file, it results in an error.

![import_error](https://github.com/user-attachments/assets/e76354c5-4b99-4394-8050-37c96ea124b0)

So, it seems that I need to rename and reorder the CSV columns to match the expected category fields.

Doing this for all categories is rather impractical.

**Reply by FlorianHeigl** [reply]
yeah that is an issue that also drove me mad

**Comment by PieterL75** [comment]
you can always create your own export templates that create the thing you need.
But keep in mind, that restoring will be very difficult. The relations to objects rely on the ID; and when you import, the ID will most likely be different.
Database backup/restore is the best way

**Comment by diasdmhub** [comment]
With the comments above, we can see how to backup Netbox DB from a Docker installation.

Basically,

```bash
docker exec -t netbox-docker-postgres-1 pg_dump -d netbox -U netbox | gzip > pg_dump.sql.gz
```

However, it is not shown how to restore such a backup.

When Netbox is started for the first time from a docker compose setup, the database is created automatically, and I assume with all the required tables. So, restoring a PostgreSQL dump, is not that simple.

Could anyone share how this process is done?

**Comment by FlorianHeigl** [comment]
Right, it had not occurred to me to update the gist with the restore steps.
But I did use the script and do restores during some horror migration. I'll
go check the shell for it and update.
... nothing left in the history because i had used screen and on that host the bash history was at default settings, so it wiped itself.
will never understand how people want defaults like that. ffs.

I wrote notes though, I'll put them here:


### DB Management / Restore 

used after major version Upgrades

Explanations regarding DB Drops - see here:
https://stackoverflow.com/questions/22287914/pg-dump-and-pg-restore-drop-if-exists

#### Docker Volume
- To reinitialize and be able to run migrations etc. I had to delete the DB Data volume
If there's no backup, and you're stuck, or whatever,
switch back to the old major version and run your backup
datenvolume loeschen (falls kein Backup, erst eine Version zurueck und sichern!)

#### Bring up and prepare DB container

IIRC this can also be done after a start of everything. You really need to drop the other containers though.
Reason being otherwise the netbox/django expectations and DB contents will be diverged a bit. So just don't start.
Update your env, pull the next version(s)

- Then Only launch DB Container
- Then initialize DB Container aufsetzen


You can use docker cp to drop the backup in the container, here I had put it under /tmp
Might need to consider access rights since the container is 'rootless'.

```
fe3a50890d8f:/# dropdb -U netbox netbox
fe3a50890d8f:/# createdb -U netbox netbox
fe3a50890d8f:/# pg_restore -U netbox -d netbox -W </tmp/db-202503091553
```
You can also only use `psql` but there's a but more crap with passing the user / password, so I didn't do that.
You can likely also use a custom initializer to load the db dump. I have not tried. I was too angry with the process.
You can also (if this is a corp setup with actual operability concerns) use a different postgres container, some can load a .sql dump on start automatically.

#### Start

After this you can start the whole netbox, and you'll see the migrations running automatically

```
netbox-1               |   Applying vpn.0006_charfield_null_choices... OK
netbox-1               |   Applying vpn.0007_natural_ordering... OK
netbox-1               |   Applying wireless.0009_wirelesslink_distance... OK
netbox-1               |   Applying wireless.0010_charfield_null_choices... OK
netbox-1               |   Applying wireless.0011_wirelesslan__location_wirelesslan__region_and_more... OK
netbox-1               |   Applying wireless.0012_alter_wirelesslan__location_and_more... OK
netbox-1               |   Applying wireless.0013_natural_ordering... OK
```

Migration issues will occur if:
- you jumped too far in versions (don't, it is not resilient and will destroy itself, but don't worry, you can do this over and over)
- you didn't properly wipe the database and there's a version issue (i.e. you got something in there that already fixed an issue a migration should handle)
- venv vs non-venv plugin installs don't match the expectations, so around the borders of pre 4.0 and after 4.0.

Generally, not all versions ran migrations/initializers automatically and you'll have a ton of issues with those.
Try to disable plugins at that point, see that you get the restore down OK, then worry about the plugins.
Sometimes you even wanna finish updating to latest and then reinstall the accompanying more current plugin versions.

__best of luck, make backups and never start to think it won't work at the end!__

**Reply by FlorianHeigl** [reply]
Oh and there's one more thing I see in the notes, you have to watch out like mad for the versions + tags.
This only matters if you are upgrading a stale env that is outdated multiple ways. You'll need to match up the compatible versions of netbox-docker and netbox, and follow through with care so you don't break things i.e. by missing an apache-nginx change etc.

## Info on Tags

That info is found here:
https://github.com/netbox-community/netbox-docker?tab=readme-ov-file#container-image-tags

(But the page has no matrix, no actual examples, and no working links)
You can look for them here: https://hub.docker.com/r/netboxcommunity/netbox/tags

Then you need to set the git branch 

git branch (i.e. `git checkout 3.2.0`, reflected in netbox-docker/VERSION)

`export VERSION=v4.2.3` (oder 4.2 to just pick **latest** - i have not tried that

`Dockerfile-Plugins: FROM netboxcommunity/netbox:v4.2.3-3.2.0`

Dazu brtaucht man ein passendes `docker-compose-override.yml`, was die build instruktionen hat.

`docker compose build --no-cache`

**Reply by FlorianHeigl** [reply]
replying put me in a time hole and now i've missed getting dinner, so this better helps someone!

**Comment by diasdmhub** [comment]
After a clear head and a bit more research, here is what I did.

### Backup

```bash
docker exec netbox-docker-postgres-1 pg_dump -Fc --clean -d netbox -U netbox | gzip > pg_dump.sql.gz
```

### Restore

> **After starting a new Netbox environment using Docker Compose (`docker compose up -d`)**

```bash
docker compose stop
docker start netbox-docker-postgres-1
docker exec netbox-docker-postgres-1 dropdb -U netbox netbox
docker exec netbox-docker-postgres-1 createdb -U netbox netbox
zcat pg_dump.sql.gz | docker exec -i netbox-docker-postgres-1 pg_restore -v --clean -d netbox -U netbox
docker exec netbox-docker-netbox-1 python3 manage.py migrate
```

---

<BR>

For the most part, it appears that all objects have been restored, except that the `Contacts` tab no longer  opens.

![contacts_tab](https://github.com/user-attachments/assets/55823c3a-328c-48a1-9bfa-930e9eb720ab)

Instead of clicking it, if I type its URL, an error shows up.

![server_error](https://github.com/user-attachments/assets/7db33e08-1f30-402c-818a-5166025fce5e)

> _The complete exception is provided below:_

```python
<class 'django.db.utils.ProgrammingError'>

column tenancy_contact.group_id does not exist
LINE 1: ...act"."description", "tenancy_contact"."comments", "tenancy_c...
                                                             ^

Python version: 3.12.3
NetBox version: 4.2.9-Docker-3.2.1
Plugins: None installed
```

The `tenancy_contact.group_id` does not exist, indeed, but why is `django.db.utils` trying to access it?

<BR>

I've tried running the "_migrate_" command, but there are no migrations to apply.

```
🧬 loaded config '/etc/netbox/config/configuration.py'
🧬 loaded config '/etc/netbox/config/extra.py'
🧬 loaded config '/etc/netbox/config/logging.py'
🧬 loaded config '/etc/netbox/config/plugins.py'
Operations to perform:
  Apply all migrations: account, auth, circuits, contenttypes, core, dcim, django_rq, extras, ipam, sessions, social_django, taggit, tenancy, users, virtualization, vpn, wireless
Running migrations:
  No migrations to apply.
```

I'm not sure what the error is. Any thoughts?

**Reply by FlorianHeigl** [reply]
can't remember. make a second install of 4.2.9-Docker-3.2.1 on the side and compare the contact/contactgroup tabels. 
I think there was some warnings regarding contactgroups in the release notes for the docker builds.

**Reply by diasdmhub** [reply]
After cloning the repository and downloading the latest image, the `contacts` menu started to work again with the correct data.

Thanks guys.
